proof/refine/RISCV64/IpcCancel_R.thy

```diff
@@ -2036,108 +3098,189 @@ proof -
+     apply (wp hoare_drop_imp)
+    apply (wpsimp wp: hoare_vcg_const_Ball_lift hoare_vcg_all_lift gts_wp' hoare_vcg_imp_lift
+                      replyUnlink_valid_objs' replyUnlink_st_tcb_at'
+                simp: valid_tcb_state'_def)+
+  apply (rule conjI)
+   apply (fastforce simp: global'_no_ex_cap pred_tcb_at'_def obj_at'_def)
   apply clarsimp
-  apply (intro conjI)
-  apply (clarsimp simp: valid_tcb_state'_def global'_no_ex_cap
-                 elim!: rsubst[where P=sym_refs]
-                 dest!: set_mono_suffix
-                intro!: ext
-       | (drule (1) bspec, clarsimp simp: valid_pspace'_def valid_tcb'_def elim!: valid_objs_valid_tcbE))+
-  done
-
-lemma ep_q_refs_max:
-  "\<lbrakk> ko_at' r p s; sym_refs (state_refs_of' s); r \<noteq> IdleEP \<rbrakk>
-      \<Longrightarrow> (state_refs_of' s p \<subseteq> (set (epQueue r) \<times> {EPSend, EPRecv}))
-       \<and> (\<forall>x\<in>set (epQueue r). \<exists>ntfnptr. state_refs_of' s x \<subseteq>
-                                  {(p, TCBBlockedSend), (p, TCBBlockedRecv), (ntfnptr, TCBBound)})"
-  apply (frule(1) sym_refs_ko_atD')
-  apply (drule ko_at_state_refs_ofD')
-  apply (case_tac r)
-    apply (clarsimp simp: st_tcb_at_refs_of_rev' tcb_bound_refs'_def
-             | rule conjI | drule(1) bspec | drule st_tcb_at_state_refs_ofD'
-             | case_tac ntfnptr)+
-  done
-
-lemma rescheduleRequired_invs'[wp]:
-  "\<lbrace>invs'\<rbrace> rescheduleRequired \<lbrace>\<lambda>rv. invs'\<rbrace>"
-  apply (simp add: rescheduleRequired_def)
-  apply (wpsimp wp: ssa_invs')
-  apply (clarsimp simp: invs'_def valid_state'_def)
+  apply (apply_conjunct \<open>intro impI\<close>,
+         (frule (1) valid_replies'_other_state; clarsimp))+
+  apply (fastforce simp: global'_no_ex_cap)
   done
 
-lemma invs_rct_ct_activatable':
-  "\<lbrakk> invs' s; ksSchedulerAction s = ResumeCurrentThread \<rbrakk>
-   \<Longrightarrow> st_tcb_at' activatable' (ksCurThread s) s"
-  by (simp add: invs'_def valid_state'_def ct_in_state'_def)
-
 lemma not_in_epQueue:
   assumes ko_at:  "ko_at' r ep_ptr s" and
           srefs:  "sym_refs (state_refs_of' s)" and
           nidle:  "r \<noteq> IdleEP" and
           st_act: "st_tcb_at' simple' t s"
   shows   "t \<notin> set (epQueue r)"
-proof
-  assume t_epQ: "t \<in> set (epQueue r)"
-
-  with ko_at nidle
-  have "(t, EPRecv) \<in> state_refs_of' s ep_ptr
-        \<or> (t, EPSend) \<in> state_refs_of' s ep_ptr"
-    by - (drule ko_at_state_refs_ofD', case_tac r, (clarsimp)+)
-
-  with ko_at srefs
-  have "(ep_ptr, TCBBlockedRecv) \<in> state_refs_of' s t
-         \<or> (ep_ptr, TCBBlockedSend) \<in> state_refs_of' s t"
-    apply -
-    apply (frule(1) sym_refs_ko_atD')
-    apply (drule ko_at_state_refs_ofD')
-    apply (case_tac r)
-      apply (clarsimp simp: st_tcb_at_refs_of_rev'
-             | drule(1) bspec | drule st_tcb_at_state_refs_ofD')+
-    done
+  proof
+    assume t_epQ: "t \<in> set (epQueue r)"
```

Comment:
The indent here looks off, `proof` should be on the same indent as the corresponding statement (lemma here), i.e. 0. Auto-indent should fix this automatically.